### ISSUE TITLE
scanning fails across ranges

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#494]
On a fresh single-node cockroach cluster:

```
~ ./cockroach put a 1 b 2 c 3 d 4
~ ./cockroach scan
"a" 1
"b" 2
"c" 3
"d" 4
~ ./cockroach split-range c c
~ ./cockroach ls-ranges
""-"c" [1]
    0: node-id=1 store-id=1 attrs=[]
"c"-"\xff\xff" [2]
    0: node-id=1 store-id=1 attrs=[]
~ ./cockroach scan a c
"a" 1
"b" 2
~./cockroach scan a d
I0327 15:38:59.262961   97373 kv.go:125] failed Scan: key range "\"a\""-"\"d\"" could not be located within a range on store
scan failed: key range "\"a\""-"\"d\"" could not be located within a range on store
```

Jira issue: CRDB-64680

**Comment by tschottdorf** [comment]
Looks like a range descriptor caching problem. See `dist_sender.go:346`:

```
// Looks up the range for [a,d) but returns stale info, so after split we still get the
// old descriptor with EndKey \xff\xff.desc, err := ds.rangeCache.LookupRangeDescriptor(args.Header().Key)
if err == nil {
	// If the request accesses keys beyond the end of this range,
	// get the descriptor of the adjacent range to address next.
	// since we get the cache, EndKey=\xff\xff and we don't enter the if clause.
	if desc.EndKey.Less(call.Args.Header().EndKey) {
		// now the request is spread out over the affected ranges, but we never get here
```

We need to invalidate the cache for that descriptor if we get a `RangeKeyMismatchError`. There could be some subtleties involved and it's likely that we have issues like this in other places...

**Comment by tschottdorf** [comment]
Ok, there is the evict logic below but it doesn't seem to be activated. Looks like the `ds.sendRPC` comes back with `err==nil` but the `reply` contains the error. So maybe the error isn't returned correctly. I'm taking a look.

**Comment by tschottdorf** [comment]
tracked this further to `rpc/send.go#sendOne`, where `call.Error == nil` holds even though the call contains an error:
```
I0327 22:27:37.929759   28373 send.go:231] returning successful call &{Node.Scan header:<timestamp:<wall_time:0 logical:0 > cmd_id:<wall_time:0 random:0 > key:"a" end_key:"d" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:0 > max_results:1000  header:<error:<range_key_mismatch:<request_start_key:"a" request_end_key:"d" > > timestamp:<wall_time:0 logical:0 > >  <nil> 0xc2084dc480} with call.Error = %!s(<nil>)
```

**Comment by tschottdorf** [comment]
after adding
```
diff --git a/kv/dist_sender.go b/kv/dist_sender.go
index 8e22958..c32a56b 100644
--- a/kv/dist_sender.go
+++ b/kv/dist_sender.go
@@ -377,6 +377,9 @@ func (ds *DistSender) Send(call *client.Call) {
                                }
                                err = ds.sendRPC(desc, call.Method, args, reply)
                        }
+                       if err == nil {
+                               err = reply.Header().GoError()
+                       }

                        if err != nil {
                                log.Warningf("failed to invoke %s: %s", call.Method, err)
```

the error is caught correctly, the cached descriptor evicted and the request re-executed. However, even though an immediate retry is announced, there's no retry and the same error is returned on the client.
The next request then succeeds because the stale descriptor is gone:

```
W0327 23:11:39.535201    9521 dist_sender.go:385] failed to invoke Scan: key range "\"a\""-"\"e\"" could not be located within a range on store
I0327 23:11:39.535282    9521 dist_sender.go:392] evicting
I0327 23:11:39.535297    9521 retry.go:93] routing Scan rpc failed; retrying immediately
I0327 23:11:45.981381    9521 txn_coord_sender.go:307] Scan: auto-wrapping in
txn and re-executing
```

**Comment by tschottdorf** [comment]
Ok, the need to retry the request from scratch is because `call.Reply` gets cross-polluted between retries. After clearing the stale descriptor, the next retry gets the same `call.Reply` which still has the range mismatch error set. On top of that we get an `OpRequiresTxnError` (which should cause an upstream retry in the `txn_coord_sender`, but since the first error is still there the `GoError()` method returns that instead (oops).
We need to properly reset `call.Reply` (I assume we don't want to replace it by a new one since upstream functions may hold on to the old ones and that could cause bugs).
I'll try to take a look this weekend, but if anyone wants to deal with this by all means go for it.

**Comment by spencerkimball** [comment]
Btw, I believe this is fixed in #450. Just FYI, Tobias that you should keep that PR in mind.

**Comment by tschottdorf** [comment]
yep, fixed in #450.
